src/notificationmanager.cpp

```diff
@@ -69,8 +70,50 @@ NotificationManager::NotificationManager(QObject* parent)
 
 NotificationManager::~NotificationManager()
 {
+    qDeleteAll(interfaces.values());
     qDeleteAll(m_Groups);
     qDeleteAll(m_unresolvedEvents);
+
+}
+
+void NotificationManager::addModem(QString path)
+{
+    DEBUG() << "NotificationManager::addModem" << path;
+    QOfonoMessageWaiting *mw = new QOfonoMessageWaiting(this);
+    interfaces.insert(path, mw);
+
+    if (!m_messageWaiting) ///TODO: hack to use first added modem, update later to support multi-modems/sims
+	    m_messageWaiting = mw;
+
+    mw->setModemPath(path);
+
+    connect(mw, SIGNAL(voicemailWaitingChanged(bool)), SLOT(slotVoicemailWaitingChanged()));
+    connect(mw, SIGNAL(voicemailMessageCountChanged(int)), SLOT(slotVoicemailWaitingChanged()));
+    connect(mw, SIGNAL(validChanged(bool)), this, SLOT(slotValidChanged(bool)));
+
+    if (mw->isValid()) {
+        DEBUG() << "QOfonoMessageWaiting already valid";
+        slotVoicemailWaitingChanged();
+    }
+}
+
+void NotificationManager::slotModemAdded(QString path)
+{
+    DEBUG() << "NotificationManager::slotModemAdded: " << path;
+    delete interfaces.take(path);
+    addModem(path);
+}
+
+void NotificationManager::slotModemRemoved(QString path)
+{
+    DEBUG() << "QOfonoMessageWaiting slotModemRemoved: " << path;
+    delete interfaces.take(path);
+}
+
+void NotificationManager::slotValidChanged(bool valid)
+{
+    DEBUG() << "QOfonoMessageWaiting slotValidChanged for some modem: " << valid;
+    slotVoicemailWaitingChanged();
```

Comment:
I think `slotVoicemailWaitingChanged()` should be invoked here only if valid is true.